src/application.rs

```diff
@@ -62,14 +67,20 @@ impl Application {
         let mut event_queue = EventQueue::new();
 
         let sender = event_queue.sender().clone(); // Collect network events
-        let network = NetworkManager::new(move |net_event| sender.send(Event::Network(net_event)));
+        let network = NetworkManager::new(move |net_event| {
+            sender.send_with_priority(Event::Network(net_event))
```

Comment:
Yes the sending events are too fast , without this termchat does not respond to term and network events till its finished

Reply:
You are right, the `ReadFile` events saturate the queue. 

What about sending the new chunk only when the previous has been processed? In this way, the events will be dispatched without fill the `termchat` memory application, only on demand. Otherwise, if the file is huge, termchat can use all the computer memory easily.

In fact, my issue that blocks termchat while is sending a file, fills the computer memory. Maybe is related.

Reply:
The chunks are processed immediately so I doubt that that would change anything.

Its actually weird that termchat gets filled with the file memory since its reading it in chunks, I'll have to investigate that.

Reply:
The thread that reads the file could be faster than the main thread that reads the event and sends the message?

Reply:
In that case, the event_queue is filled with chunks

Reply:
That actually makes a lot of sense.

But how to handle this, do I have to rate limit reading somehow?

Reply:
@lemunozm can you test this branch https://github.com/lemunozm/termchat/commit/7043901332f88f5caebbcbce6b2dbd40b7649127

This should fix the sending, if this idea work I'll work on the receiving later

Reply:
Actually this appears to be do the trick https://github.com/sigmaSd/termchat/tree/park

I sent a 200mb file without the receiver nor the sender exceeding more than 1mb usage.

Also now this works without send_with_priority

Let me know what you think

Reply:
> But how to handle this, do I have to rate limit reading somehow?

The file should be read a new chunk once the previous chunk has been sent. 
I will see your branches during the day :) Thanks!

Reply:
Thats what the new branch doe

Reply:
Hi @sigmaSd, `termchat` continues blocking while send file :(

Commit: "d15fa5bc" (your park branch), in a clean state. The issue happens even if the file is not found and even if is the only termchat in the network. I have to debug it to clarify something. Maybe some thread is blocked on my computer for some unknown reason.

Regarding the approach, I think that there is no need to create more threads than the main one if once the file is open, the descriptor is stored in the application. Each new ReadFile event will send the chunk and will read more data and enqueue other ReadFile event, and so on.

Reply:
I want to add a `file-trasnfer` example to message-io because I think that this use case could be really useful

Reply:
I tried to remove the threads https://github.com/sigmaSd/termchat/tree/without_threads but the problem is the event queue will be filled up with chunks again and I cant use thread::park trick anymore to throttle the reading.

Otherwise I think threads should be ok , the only problem is the issue you're having, maybe you gdb -p (pidof termchat) and use backtrace to see where its hanging.

I'll also see your message-io sending file example for more ideas.

Reply:
Also the above method that I tried blocks everything.

I think your idea to save  file descriptors means I'll have to add at least one new thread that polls all the new files, instead of the park method that creates a thread per file sent, If that is the case, I think the advantage is not that important

Reply:
Regarding your issue, by blocking you mean that you see the progress bar filling but you can't input keys and receive network events till its done?

If that's the case, the only thing I can think of is that std::thread::park is not being respected , it does say on the readme that the thread can wake up even with park https://doc.rust-lang.org/std/thread/fn.park.html .But that would be surprising.

I'll just for sanity push a branch that uses mutex synchronisation instead, so you can test it

Reply:
@lemunozm here is a branch with condvar/mutex synchro which should be immune to spurious wake-ups. https://github.com/sigmaSd/termchat/tree/condvar

Reply:
Thanks for your investigation! I have finished the `file-transfer` example in branch: https://github.com/lemunozm/message-io/tree/example/file_transfer .

It works, but randomly, with long files (>100M) the receiver stops receiving. It is possible that the issue is related to some network saturation that message-io is not handling correctly (the socket is in would block state when it happens and never recovery), and is possible to be related with the `termthat` issues.

> I think your idea to save file descriptors means I'll have to add at least one new thread that polls all the new files, instead of the park method that creates a thread per file sent, If that is the case, I think the advantage is not that important

Not exactly, the branch I have upload show it :)

> Regarding your issue, by blocking you mean that you see the progress bar filling but you can't input keys and receive network events till its done?

Input blocked (ctrl-c doesn't works for example), and no command or progress bar is printed on the screen of the sender. The command is printed in the receiver and this termchat continues working.

I will test your `condvar` branch in a while, thanks!

Reply:
Your example is clever! I implemented the idea https://github.com/sigmaSd/termchat/tree/clever_sync maybe this will work without issues.

Reply:
There is no much error handling like in termchat but the example focus on to the point :)

I was testing the message-io example issue and I pretty sure that is not related to the network/threads. It is an encoding problem. It seems to be related when a socket read data that contains more than one message at once. I thought that I was taking that into account but seems to be buggy. Randomly, the receiver is a little bit slower than the sender, and when you try to read data, there is more than a message. At that moment, the encoding header loses and the decoder thinks that the message has infinity size. Not sending any more packets.

Maybe the termchat problem is related, and my computer crashes because of computer timings or how the OS join the packets

Reply:
Well till that's fixed I'm probably going to merge https://github.com/sigmaSd/termchat/tree/clever_sync to this pr. It looks correct and it works on my setup.

Do you have some remark regarding it?

Also do you have a test case for the issue you're having with message-io? I'm actually currently not able to reproduce the issue, but It happened sporadically before and your diagnosis aligns with what I experienced.

Reply:
Hi @sigmaSd , Yes, I made a UT that reproduce the encoding issue and also I merged the fix along with 0.4.5 version to `message-io`. Your branch should work fine with this version :)

The `clever_sync` branch looks good! 
I would write some minor comment when you merge it to this PR

Thanks!

Reply:
Merged + updated to the new message-io, Hopefully its fixed now!